src/iop/overlay.c

```diff
@@ -250,8 +250,7 @@ static void _clear_cache_entry(dt_iop_module_t *self, const int index)
 }
 
 static void _module_remove_callback(gpointer instance,
-                                    dt_iop_module_t *self,
-                                    gpointer user_data)
+                                    dt_iop_module_t *self)
```

Comment:
Would you comment on this ? We reduce number of parameters - superfluous?

Reply:
overzealous :-)

According to the signal signature, the `user_data` parameter should be there. But since it (currently) isn't used, there is no impact.

tbh don't know what I was thinking (if anything) when I edited this, except that in this case the "self" parameter is confusing, since it is actually the first parameter passed from `develop.c` when it throws the signal `DT_SIGNAL_DEVELOP_MODULE_REMOVE` and the `user_data` parameter is the real `self` (i.e. the overlay module) set by `DT_DEBUG_CONTROL_SIGNAL_CONNECT`.

So, I may be completely mistaken, but it looks to me that this handler tries to interpret _any_ module being removed as if it was an overlay and looks up the `imgid` field (which would mean something completely different in another module's blob) and if it _happens_ to be a valid id and used in an overlay, it will be `dt_overlay_remove`d (which might be bad?). If however this _is_ the current module being removed, then it probably does the right thing. I imagine there will be a bug when there is more than one overlay module and only one of them is removed, but haven't tested because I don't know what the impact of `dt_overlay_remove` is supposed to be.
 @turbogit ?

What doesn't help here is that the definition of in `DT_SIGNAL_DEVELOP_MODULE_REMOVE` in `signal.c` is actually marked incorrectly with `DT_SIGNAL_MODULE_REMOVE`, so not easy to find. This is fixed in 85956e13ada6fe20eccca0feed7de45a0fbab4a8

Reply:
In intensive care a overzealous doc is likely harming :-)

As this signal is `special` to this module i think self/user_data will **always** be a ptr to an overlay ...

Yes i have seen (and tried the signal fixing commit). This PR is by far large enough to be reviewed. I looked at my watch while enjoying and keeping concentrated , 2.5 hrs :-)

Reply:
btw bit confused by your comment in PR additional fixes commit "A few places where i think we should keep the checks for allocations even if allocations errors are very unlikely due to requested size." since I don't see those removals in my "local" PR. Maybe they were introduced by rebasing/cherry-picking (which is always tricky/dangerous)?

You did a very good read-through to catch all the other fixes!

Reply:
> In intensive care a overzealous doc is likely harming :-)

You are absolutely right and that's exactly how I meant it! ;-)

> As this signal is `special` to this module

It isn't? It is sent to `history.c` for _any_ module being removed.

> i think self/user_data will **always** be a ptr to an overlay ...

`user_data` yes, but not the second ("self") pointer. I suggest changing this to 
```
static void _module_remove_callback(gpointer instance,
                                    dt_iop_module_t *removed,
                                    dt_iop_module_t *self)
```
and fixing the code accordingly. Probably by adding a check `removed == self`, but that's up to @turbogit

Reply:
Isn't overlay the only module asking to receive it - at least in current code?

Reply:
```
$ grep CONNECT.*DT_SIGNAL_DEVELOP_MODULE_REMOVE src -R
src/iop/overlay.c:  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_DEVELOP_MODULE_REMOVE,
src/libs/history.c:  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_DEVELOP_MODULE_REMOVE,
```
Those 2.5hrs must have been exhausting... ;-)

Reply:
Well your work certainly took longer... It's just if I push that PR I wanted to be certain about what's happening. 

Will check the overlay code an propose a correct solution.

Reply:
Nice, Pascal has merged that big first part....